src/libmongoc/tests/test-mongoc-hello.c

```diff
@@ -0,0 +1,200 @@
+   ASSERT (client);
+
+   ping = BCON_NEW ("ping", BCON_INT32 (1));
+
+   // Use a "future" function to send a ping command in the background. 
+   future = future_client_command_simple (
+      client, "db", ping, NULL, NULL, &error);
+
+   /* Since this is the first command, a new connection is opened. */
+   if(-1 != requested_server_api_version) {
+   	request = mock_server_receives_hello(server);
+   }
+   else {
+	// legacy API:
+   	request = mock_server_receives_legacy_hello (server, "{'isMaster': 1}");
+   } 
+   ASSERT (request);
+
+   // Note that this basically works against the mock server as there's no "real" legacy mode:
+   mock_server_replies_simple (
+      request, "{'ok': 1, 'isWritablePrimary': true, 'maxWireVersion': 14 }");
+   request_destroy (request);
+
+   // Now expect the ping command that we launched earlier:
+   request = mock_server_receives_msg (
+      server, MONGOC_MSG_NONE, tmp_bson ("{'ping': 1}"));
+   ASSERT (request);
+   mock_server_replies_simple (request,
+                               "{'ok': 1, 'isWritablePrimary': true }");
+   request_destroy (request);
+
+   ret = future_get_bool (future);
+   ASSERT (ret);
+
+   // Tidy up:
+   future_destroy (future);
+   bson_destroy (ping);
+   mongoc_client_destroy (client);
+   mock_server_destroy (server);
+
+   /* If we made it here, everything went just fine: */
+   return true;
+}
+
+void
+test_mongoc_hello ()
+{
+   if (!TestSuite_CheckMockServerAllowed ()) {
+      return;
+   }
+
+   // Always check with the default protocol version (which here may come from the
+   // environment; this is "legacy hello" in this test):
+   _test_mongoc_hello_impl (-1);
+
+   // Check with non-legacy hello:
+   _test_mongoc_hello_impl (MONGOC_SERVER_API_V1);
+}
+
+bool
+_test_mongoc_hello_client_pool_impl (int requested_server_api_version)
+{
+   mock_server_t *server;
+   mongoc_uri_t *test_uri;
+   mongoc_client_pool_t *pool;
+   mongoc_client_t *client; 
+   request_t *request;
+
+   mongoc_server_api_t *requested_server_api = NULL;
+
+   if (-1 != requested_server_api_version) {
+      requested_server_api =
+         mongoc_server_api_new (requested_server_api_version);
+      ASSERT (requested_server_api);
+   }
+
+   MONGOC_DEBUG (
+      "using requested_server_api_version == %d; requested_server_api = %p\n",
+      requested_server_api_version,
+      (void *) requested_server_api);
+
+   server = mock_server_new ();
+   mock_server_run (server);
+
+   test_uri = mongoc_uri_copy (mock_server_get_uri (server));
+
+   pool =
+      test_framework_client_pool_new_from_uri (test_uri, requested_server_api);
+   BSON_ASSERT (pool);
+
+   test_framework_set_pool_ssl_opts (pool);
+
+   client = mongoc_client_pool_pop (pool);
+   ASSERT(client);
+
+   // Popping the client should have triggered a connection:
+   if(-1 != requested_server_api_version) {
+        // a specific API version has been requested:
+   	request = mock_server_receives_hello(server);
```

Comment:
I think these tests pass without additional changes behavior changes.

`mock_server_receives_hello` uses `mock_server_receives_command`.
`mock_server_receives_command` expects OP_QUERY.

I suggest taking a step back. Start on a fresh branch.

The proposed behavior change is to use OP_MSG for the first hello if a server API version is requested. Try the red/green/refactor method:

1. Add failing tests.
2. Make the behavior change to get passing tests.
3. Refactor as needed.